embassy-nrf/src/usb.rs

```diff
@@ -1,43 +1,798 @@
+                    w.ep().bits(ep_addr.index() as u8 & 0b111);
+                    w.io().bit(ep_addr.is_in());
+                    w.stall().bit(stalled)
+                });
+            }
+        }
+
+        //if stalled {
+        //    self.busy_in_endpoints &= !(1 << ep_addr.index());
+        //}
+    }
+
+    fn is_stalled(ep_addr: EndpointAddress) -> bool {
+        let regs = T::regs();
+
+        let i = ep_addr.index();
+        match ep_addr.direction() {
+            UsbDirection::Out => regs.halted.epout[i].read().getstatus().is_halted(),
+            UsbDirection::In => regs.halted.epin[i].read().getstatus().is_halted(),
+        }
+    }
 }
 
-impl<'d, T: Instance> UsbBus<'d, T> {
-    pub fn new(_usb: impl Unborrow<Target = T> + 'd) -> UsbBusAllocator<Usbd<UsbBus<'d, T>>> {
-        let r = T::regs();
+impl<'d, T: Instance> driver::Driver<'d> for Driver<'d, T> {
+    type EndpointOut = Endpoint<'d, T, Out>;
+    type EndpointIn = Endpoint<'d, T, In>;
+    type ControlPipe = ControlPipe<'d, T>;
+    type Bus = Bus<'d, T>;
+
+    fn alloc_endpoint_in(
+        &mut self,
+        ep_addr: Option<EndpointAddress>,
+        ep_type: EndpointType,
+        max_packet_size: u16,
+        interval: u8,
+    ) -> Result<Self::EndpointIn, driver::EndpointAllocError> {
+        let index = self
+            .alloc_in
+            .allocate(ep_addr, ep_type, max_packet_size, interval)?;
+        let ep_addr = EndpointAddress::from_parts(index, UsbDirection::In);
+        Ok(Endpoint::new(EndpointInfo {
+            addr: ep_addr,
+            ep_type,
+            max_packet_size,
+            interval,
+        }))
+    }
+
+    fn alloc_endpoint_out(
+        &mut self,
+        ep_addr: Option<EndpointAddress>,
+        ep_type: EndpointType,
+        max_packet_size: u16,
+        interval: u8,
+    ) -> Result<Self::EndpointOut, driver::EndpointAllocError> {
+        let index = self
+            .alloc_out
+            .allocate(ep_addr, ep_type, max_packet_size, interval)?;
+        let ep_addr = EndpointAddress::from_parts(index, UsbDirection::Out);
+        Ok(Endpoint::new(EndpointInfo {
+            addr: ep_addr,
+            ep_type,
+            max_packet_size,
+            interval,
+        }))
+    }
+
+    fn alloc_control_pipe(
+        &mut self,
+        max_packet_size: u16,
+    ) -> Result<Self::ControlPipe, driver::EndpointAllocError> {
+        self.alloc_endpoint_out(Some(0x00.into()), EndpointType::Control, max_packet_size, 0)?;
+        self.alloc_endpoint_in(Some(0x80.into()), EndpointType::Control, max_packet_size, 0)?;
+        Ok(ControlPipe {
+            _phantom: PhantomData,
+            max_packet_size,
+        })
+    }
+
+    fn enable(self) -> Self::Bus {
+        let regs = T::regs();
+
+        errata::pre_enable();
+
+        regs.enable.write(|w| w.enable().enabled());
+
+        // Wait until the peripheral is ready.
+        while !regs.eventcause.read().ready().is_ready() {}
```

Comment:
The datasheet indicates this will typically take 1ms from enable to ready. That seems like a long time to block the executor. Should `Driver::enable` be async? That would require `UsbDriverBuilder::build` to become async as well.

Also, should `Bus` disable the `USBD` on drop?

Reply:
> That seems like a long time to block the executor. Should Driver::enable be async?

If it's only at startup maybe it's not a big deal. Not sure if there's any usecase where you'd want to create/drop it dynamically at runtime? No strong opinion, feel free to make it async if you want.

> Also, should Bus disable the USBD on drop?

Ideally yes. I think due to how the lifetimes are setup it's possible for individual endpoints to outlive the Bus, that might be tricky to handle.